### PR TITLE
Powershell fixes for cross-subscription

### PR DESCRIPTION
**Pull request by singhabh27**
## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from the scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** change version of module in pull request

### CODE REVIEW

File: src/Migrate/custom/New-AzMigrateServerReplication.ps1 [hunk 1]
```diff
@@ -617,8 +617,8 @@ public static int hashForArtifact(String artifact)
         }
 
         Import-Module Az.Resources
-        $TargetResourceGroupName = $ProviderSpecificDetails.TargetResourceGroupId.Split('/')[4]
-        $VMNamePresentinRg = Get-AzResource -ResourceGroupName $TargetResourceGroupName -Name $TargetVMName -ResourceType "Microsoft.Compute/virtualMachines" -ErrorVariable notPresent -ErrorAction SilentlyContinue
+        $vmID = $ProviderSpecificDetails.TargetResourceGroupId + "/providers/Microsoft.Compute/virtualMachines/" + $TargetVMName
```
Comment (Arpitaji) [inline src/Migrate/custom/New-AzMigrateServerReplication.ps1:620]: >$vmID [](http://example.com/codeflow?start=8&length=5)

nit $vmId

**Comment by Arpitaji** [comment]
>             $TargetResourceGroupName = $ProviderSpecificDetails.TargetResourceGroupId.Split('/')[4]

Please remove

---
Refers to: src/Migrate/custom/Set-AzMigrateServerReplication.ps1:509 in 98adb92. [](commit_id = 98adb926777fd4e76cca8bde62f82e23c3c55087, deletion_comment = False)

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 2]
```diff
@@ -512,8 +512,8 @@ function Set-AzMigrateServerReplication {
                 if ($TargetVMName.length -gt 64 -or $TargetVMName.length -eq 0) {
                     throw "The target virtual machine name must be between 1 and 64 characters long."
                 }
-
-                $VMNamePresentinRg = Get-AzResource -ResourceGroupName $TargetResourceGroupName -Name $TargetVMName -ResourceType "Microsoft.Compute/virtualMachines" -ErrorVariable notPresent -ErrorAction SilentlyContinue
+                $vmID = $ProviderSpecificDetails.TargetResourceGroupId + "/providers/Microsoft.Compute/virtualMachines/" + $TargetVMName
```
Comment (Arpitaji) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:515]: >$vmID [](http://example.com/codeflow?start=16&length=5)

nit: vmId

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 3]
```diff
@@ -533,7 +533,8 @@ function Set-AzMigrateServerReplication {
                 $duplicateDiskName = New-Object System.Collections.Generic.HashSet[String]
                 $uniqueDiskUuids = [System.Collections.Generic.HashSet[String]]::new([StringComparer]::InvariantCultureIgnoreCase)
                 foreach($DiskObject in $DiskToUpdate) {
-                    $diskNamePresent = Get-AzResource -ResourceGroupName $TargetResourceGroupName -Name $DiskObject.TargetDiskName -ResourceType "Microsoft.Compute/disks" -ErrorVariable notPresent -ErrorAction SilentlyContinue
+                    $diskID = $ProviderSpecificDetails.TargetResourceGroupId + "/providers/Microsoft.Compute/disks/" + $DiskObject.TargetDiskName
```
Comment (Arpitaji) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:536]: >$diskID [](http://example.com/codeflow?start=20&length=7)

nit: diskId

File: src/Migrate/custom/Set-AzMigrateServerReplication.ps1 [hunk 4]
```diff
@@ -618,7 +621,8 @@ function Set-AzMigrateServerReplication {
                         $updateNic.TargetSubnetName = $matchingUserInputNic.TargetSubnetName
                     }
                     if ($null -ne $matchingUserInputNic.TargetNicName) {
-                        $nicNamePresent = Get-AzResource -ResourceGroupName $TargetResourceGroupName -Name $matchingUserInputNic.TargetNicName -ResourceType "Microsoft.Network/networkInterfaces" -ErrorVariable notPresent -ErrorAction SilentlyContinue
+                        $nicID = $ProviderSpecificDetails.TargetResourceGroupId + "/providers/Microsoft.Network/networkInterfaces/" + $matchingUserInputNic.TargetNicName
```
Comment (Arpitaji) [inline src/Migrate/custom/Set-AzMigrateServerReplication.ps1:624]: >$nicID [](http://example.com/codeflow?start=24&length=6)

nit: nicId

File: src/Migrate/custom/Start-AzMigrateTestMigration.ps1 [hunk 5]
```diff
@@ -131,7 +131,10 @@ function Start-AzMigrateTestMigration {
             Import-module -Name Az.Compute
             Import-module -Name Az.Network
             $AvSetName = $AvSetId.Split("/")[-1];
-            $AvSetRg = $AvSetId.Split("/")[-5];
+            $AvSetRg = $AvSetId.Split("/")[-5];            
+            $TargetSubscriptionId = $AvSetId.Split("/")[-7];
+            $SourceSubscriptionId = (Get-AzContext -ErrorVariable notPresent -ErrorAction SilentlyContinue).Subscription.Id
+            Set-AzContext -SubscriptionId $TargetSubscriptionId -ErrorVariable notPresent -ErrorAction SilentlyContinue
             $AvSet = Get-AzAvailabilitySet -ResourceGroupName $AvSetRg -Name $AvSetName -ErrorVariable notPresent -ErrorAction SilentlyContinue
```
Comment (Arpitaji) [inline src/Migrate/custom/Start-AzMigrateTestMigration.ps1:138]: >AzAvailabilitySet [](http://example.com/codeflow?start=25&length=17)

can we not use the Get-AzResource to get?

File: src/Migrate/custom/Start-AzMigrateTestMigration.ps1 [hunk 6]
```diff
@@ -161,6 +164,7 @@ function Start-AzMigrateTestMigration {
                     }
                 }
             }
+            Set-AzContext -SubscriptionId $SourceSubscriptionId -ErrorVariable notPresent -ErrorAction SilentlyContinue
```
Comment (Arpitaji) [inline src/Migrate/custom/Start-AzMigrateTestMigration.ps1:167]: >Set-AzContext [](http://example.com/codeflow?start=12&length=13)

If a failure occurs we wont be able to change context, can this cause issue?

Comment (singhabh27) [reply src/Migrate/custom/Start-AzMigrateTestMigration.ps1:167]: We need to change context to older context for failed case.

File: src/Migrate/custom/Start-AzMigrateTestMigration.ps1 (hunk 5 above)
Comment (singhabh27) [reply src/Migrate/custom/Start-AzMigrateTestMigration.ps1:138]: We are fetching VM details from Avset, which we can't do using Get-AzResource

**Review by prsadhu-ms-idc — COMMENTED** [review]
LGTM

**Review by BethanyZhou — APPROVED** [review]
Looks good to me

**Pull request merged**